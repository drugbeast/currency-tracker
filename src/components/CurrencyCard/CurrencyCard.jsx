import { useContext, useEffect, useState } from 'react'
import { createPortal } from 'react-dom'

import currencies from '../../constants/currencies'
import icons from '../../constants/icons'
import { CurrenciesContext } from '../../pages/Home/Home'
import Modal from '../Modal/Modal'
import styles from './CurrencyCard.module.scss'

function CurrencyCard() {
  const [isClicked, setClicked] = useState(false)
  const { symbol, rate } = useContext(CurrenciesContext)

  const onPress = e => {
    if (e.key === 'Escape') {
      setClicked(false)
    }
  }

  useEffect(() => {
    document.addEventListener('keydown', onPress)
    return () => document.removeEventListener('keydown', onPress)
  }, [])

  return (
    <section className={styles.card} onClick={() => setClicked(true)}>
      {isClicked && createPortal(<Modal />, document.body)}
      <div className={styles.inner}>
        {icons[symbol]({ className: styles.icon })}
        <div className={styles.info}>
          <span className={styles.title}>{currencies[symbol]}</span>
          <span className={styles.rate}>R&#36; {rate}</span>
        </div>
      </div>
    </section>
  )
}

export default CurrencyCard
