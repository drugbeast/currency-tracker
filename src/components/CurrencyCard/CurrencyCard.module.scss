@use '../../constants/constants.scss' as *;

.card {
  cursor: pointer;
  background-color: var(--theme-color-8);
  padding: 32px;
  border: 1px solid var(--theme-color-9);
  width: 42%;
  border-radius: 8px;
}
.inner {
  display: flex;
  gap: 24px;
}
.info {
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 4px;
}
.title {
  font-size: 35px;
  font-weight: 400;
  color: var(--card-text);
}
.rate {
  font-size: 32px;
  color: var(--theme-color-10);
}
.icon {
  height: 80px;
  width: 80px;
}

@media screen and (max-width: 1280px) {
  .title {
    font-size: 26px;
  }
  .rate {
    font-size: 24px;
  }
  .icon {
    height: 60px;
    width: 60px;
  }
}

@media screen and (max-width: 1024px) {
  .card {
    padding: 16px;
  }
  .title {
    font-size: 20px;
  }
  .rate {
    font-size: 20px;
  }
  .icon {
    height: 50px;
    width: 50px;
  }
}

@media screen and (max-width: 768px) {
  .card {
    width: 100%;
  }
}
