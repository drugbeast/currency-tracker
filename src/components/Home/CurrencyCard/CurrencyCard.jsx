import { CURRENCIES_SECOND_FORM } from 'Constants/constants'
import icons from 'Constants/icons'
import { memo } from 'react'

import styles from './CurrencyCard.module.scss'

const CurrencyCard = memo((props) => {
  const { setCardClicked, setShow, symbol, rate } = props
  const onCardClick = () => {
    setShow(true)
    setCardClicked({ rate: Number(rate), symbol })
  }
  return (
    <section className={styles.card} onClick={onCardClick} data-cy="card">
      <div className={styles.inner}>
        {icons[symbol]({ className: styles.icon })}
        <div className={styles.info}>
          <span className={styles.title}>{CURRENCIES_SECOND_FORM[symbol]}</span>
          <span className={styles.rate}>R&#36; {rate}</span>
        </div>
      </div>
    </section>
  )
})

export default CurrencyCard
