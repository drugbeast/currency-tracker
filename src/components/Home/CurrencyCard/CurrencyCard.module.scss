@use '../../../constants/mixins.scss' as *;
@use '../../../constants/breakpoints';

.card {
  cursor: pointer;
  background-color: var(--theme-color-8);
  padding: 32px;
  border: 1px solid var(--theme-color-9);
  width: 42%;
  border-radius: 8px;
}

.inner {
  @include flex(stretch, stretch, 24px);
}

.info {
  @include flex(center, stretch, 4px, column);
}

.title {
  font-size: 35px;
  font-weight: 400;
  color: var(--card-text);
}

.rate {
  font-size: 32px;
  color: var(--theme-color-10);
}

.icon {
  height: 80px;
  width: 80px;
}

@media screen and (max-width: breakpoints.$desktop) {
  .title {
    font-size: 26px;
  }
  .rate {
    font-size: 24px;
  }
  .icon {
    height: 60px;
    width: 60px;
  }
}

@media screen and (max-width: breakpoints.$desktop-mini) {
  .card {
    padding: 16px;
  }
  .title {
    font-size: 20px;
  }
  .rate {
    font-size: 20px;
  }
  .icon {
    height: 50px;
    width: 50px;
  }
}

@media screen and (max-width: breakpoints.$tablet) {
  .card {
    width: 100%;
  }
}
