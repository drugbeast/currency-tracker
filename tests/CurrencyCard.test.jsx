import '@testing-library/jest-dom'

import { fireEvent, render, screen } from '@testing-library/react'
import CurrencyCard from 'Components/Home/CurrencyCard/CurrencyCard'

describe('Currency card component', () => {
  const setCardClicked = jest.fn()
  const setShow = jest.fn()
  const symbols = ['AUD', 'BTC', 'CAD', 'USD', 'EUR', 'TRY', 'ARS', 'KRW', 'JPY']
  const mockCurrency = 'AUD'
  const mockRate = 1

  const renderComponent = (symbol = mockCurrency) => {
    render(
      <CurrencyCard
        setCardClicked={setCardClicked}
        setShow={setShow}
        symbol={symbol}
        rate={mockRate}
      />,
    )
  }

  it('should render the component', () => {
    renderComponent()
    expect(screen.getByRole('presentation')).toBeInTheDocument()
  })

  it('should call callbacks after card was clicked', () => {
    renderComponent()
    const card = screen.getByRole('presentation')
    fireEvent.click(card)
    expect(setCardClicked).toHaveBeenCalledTimes(1)
    expect(setShow).toHaveBeenCalledTimes(1)
  })

  it('should match snapshot', () => {
    const CurrencyCardComponent = renderComponent()
    expect(CurrencyCardComponent).toMatchSnapshot()
  })

  it(`should render cards with all currencies's svgs`, () => {
    symbols.forEach((symbol) => {
      renderComponent(symbol)
      expect(screen.getAllByRole('presentation').length).toBeGreaterThan(0)
    })
  })
})
