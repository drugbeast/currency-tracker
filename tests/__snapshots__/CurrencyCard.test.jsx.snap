// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency card component should match snapshot 1`] = `undefined`;
